community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeIntoAcceptanceTest.scala

```diff
@@ -0,0 +1,112 @@
+    val update = execute("""MATCH (a {name:'A'}), (b {name:'B'})
+      |MERGE (a)-[r:TYPE]->(b) ON CREATE SET r.name = 'foo'""".stripMargin)
+
+    //then
+    update should use("Merge(Into)")
+    val res = execute("MATCH ()-[r:TYPE]->() RETURN extract(key IN keys(r)| key + '->' + r[key]) as keyValue")
+    res.toList should equal(List(Map("keyValue" -> Seq("name->foo"))))
+  }
+
+  test("ON CREATE with deleting one property") {
+    //given
+    createNode("A")
+    createNode("B")
+
+    //when
+    val update = execute("""MATCH (a {name:'A'}), (b {name:'B'})
+                           |MERGE (a)-[r:TYPE]->(b) ON CREATE SET r.name = null""".stripMargin)
+
+    //then
+    update should use("Merge(Into)")
+    val res = execute("MATCH ()-[r:TYPE]->() RETURN extract(key IN keys(r)| key + '->' + r[key]) as keyValue")
+    res.toList should equal(List(Map("keyValue" -> Seq.empty)))
+  }
+
+  test("ON CREATE with update all properties from node") {
+    //given
+    createNode("A")
+    createNode("B")
+
+    //when
+    val update = execute("MATCH (a {name:'A'}), (b {name:'B'}) MERGE (a)-[r:TYPE]->(b) ON CREATE SET r = a")
+
+    //then
+    update should use("Merge(Into)")
+    val res = execute("MATCH ()-[r:TYPE]->() RETURN extract(key IN keys(r)| key + '->' + r[key]) as keyValue")
+    res.toList should equal(List(Map("keyValue" -> Seq("name->A"))))
+  }
+
+  test("ON MATCH with update all properties from node") {
+    //note the props here should be overwritten with ON MATCH
+    relate(createNode("A"), createNode("B"), "TYPE", Map("foo" -> "bar"))
+
+    //when
+    val update = execute("MATCH (a {name:'A'}), (b {name:'B'}) MERGE (a)-[r:TYPE]->(b) ON MATCH SET r = a")
+
+    //then
+    update should use("Merge(Into)")
+    val res = execute("MATCH ()-[r:TYPE]->() RETURN extract(key IN keys(r)| key + '->' + r[key]) as keyValue")
+    res.toList should equal(List(Map("keyValue" -> Seq("name->A"))))
+  }
+
+  test("ON CREATE with update properties from literal map") {
+    //given
+    createNode("A")
+    createNode("B")
+
+    //when
+    val update = execute("""MATCH (a {name:'A'}), (b {name:'B'})
+      |MERGE (a)-[r:TYPE]->(b) ON CREATE SET r += {foo: 'bar', bar: 'baz'}""".stripMargin)
+
+    //then
+    update should use("Merge(Into)")
+    val res = execute("MATCH ()-[r:TYPE]->() RETURN extract(key IN keys(r)| key + '->' + r[key]) as keyValue")
+    res.toList should equal(List(Map("keyValue" -> Seq("foo->bar", "bar->baz"))))
+  }
+
+
+  test("ON MATCH with update properties from literal map") {
+    //given
+    relate(createNode("A"), createNode("B"), "TYPE", Map("foo" -> "bar"))
+
+    //when
+    val update = execute("""MATCH (a {name:'A'}), (b {name:'B'})
+                           |MERGE (a)-[r:TYPE]->(b) ON MATCH SET r += {foo: 'baz', bar: 'baz'}""".stripMargin)
+
+    //then
+    update should use("Merge(Into)")
+    val res = execute("MATCH ()-[r:TYPE]->() RETURN extract(key IN keys(r)| key + '->' + r[key]) as keyValue")
```

Comment:
Yeah same as above, here I want to make sure the existing properties are not overwritten. But I do agree it is somewhat convoluted.